exporter/exporterhelper/internal/queuebatch/config.go

```
@@ -0,0 +1,100 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package queuebatch // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch"
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/exporter/exporterbatcher"
+)
+
+// Config defines configuration for queueing and batching incoming requests.
+type Config struct {
+	// Enabled indicates whether to not enqueue and batch before exporting.
+	// TODO: This will be removed when Optional is available.
+	Enabled bool `mapstructure:"enabled"`
+
+	// WaitForResult determines if incoming requests are blocked until the request is processed or not.
+	// Currently, this option is not available when persistent queue is configured using the storage configuration.
+	WaitForResult bool `mapstructure:"wait_for_result"`
+
+	// Sizer configuration determines the type of size measurement used by this component.
+	// It accepts "requests", "items", or "bytes".
+	Sizer exporterbatcher.SizerType `mapstructure:"sizer"`
```

Do you want to use one field for `queue_size` and batch sizes going forward? This might be confusing given that we have `requests` for the queue and `items` for batcher by default. I think we shouldn't change that default and split this field instead to make it clear. Probably, just expose only batch sizer until the queue sizer is available for now?

---

>Do you want to use one field for queue_size and batch sizes going forward? 

I would definitely prefer this. 

> This might be confusing given that we have requests for the queue and items for batcher by default. 

Batching is experimental and not enabled by default, so we can tune that.

> I think we shouldn't change that default and split this field instead to make it clear. 

Will make everything harder to resonate in my opinion that in the same config you have "sizes" in different parts that are using different sizers.

> Even without the default, we might want to give the flexibility to use different sizers, especially given that we won't always have all the options consistently available for both.

For queue we will have all for sure because we will implement them. For batch, yes we may have different supported (requests is always though).

> For now, we can probably just expose only the batch sizer until the queue sizer is available. 

As mentioned in the comment, this is the final config I want to get to, and will implement what is missing before then.

---

> Will make everything harder to resonate in my opinion that in the same config you have "sizes" in different parts that are using different sizers.

The goal sounds good to me. It's just unclear to me how to achieve that.

Do you want to keep the existing default of `items` for batcher and `requests` for the queue or you want to have one particular sizer by default for both?

> For queue we will have all for sure because we will implement them. For batch, yes we may have different supported (requests is always though).

What would be the behavior if user picks a sizer that is only available for the queue and enables batcher with some custom numbers there?

---

>What would be the behavior if user picks a sizer that is only available for the queue and enables batcher with some custom numbers there?

I think we should try to fail during config validation (or at most start time) if the sizer picked is not valid. I have some ideas on how to do that (for me that is implementation detail), but want to focus on how the config should look like for the user and then figure out how can I fail fast.

---

>The goal sounds good to me. It's just unclear to me how to achieve that. Do you want to keep the existing default of items for batcher and requests for the queue or you want to have one particular sizer by default for both?

I think in the end the default sizer for this component (for all sizes) should be "items", which has the best tradeoff between performance (cpu cost) and granularity.

---

> I think we should try to fail during config validation (or at most start time) if the sizer picked is not valid. I have some ideas on how to do that (for me that is implementation detail), but want to focus on how the config should look like for the user and then figure out how can I fail fast.

Ok. It means that some sizer options won't be available if users want to use both queue and batching. That sounds reasonable. We should probably implement all the sizers before migrating to this interface.

> I think in the end the default sizer for this component (for all sizes) should be "items", which has the best tradeoff between performance (cpu cost) and granularity.

Sounds reasonable to me. We need to think carefully about a migration approach here I guess.

---

Question: I wasn't sure which exporters would not support certain sizer options. Isn't it possible, for exporters that don't implement a given sizer option, to use a default implementation of that sizer? For requests and items, there is a natural definition. For bytes, when the exporter doesn't implement a specific sizer, couldn't we silently fall back on the OTLP sizer?

---

> For bytes, when the exporter doesn't implement a specific sizer, couldn't we silently fall back on the OTLP sizer?

Not really, since splitting the data may not be a valid operation after the conversion to the exporter specific format.